Modules/_bz2module.c

```diff
@@ -646,20 +687,55 @@ _bz2_BZ2Decompressor___init___impl(BZ2Decompressor *self)
     return -1;
 }
 
+static int
+_bz2_BZ2Decompressor___init__(PyObject *self, PyObject *args, PyObject *kwargs)
+{
+    int return_value = -1;
+    _bz2_state *state = PyType_GetModuleState(Py_TYPE(self));
+    assert(state != NULL);
+    
+    if (Py_IS_TYPE(self, state->bz2_decompressor_type) &&
```

Comment:
I don't think that this test makes sense.

PyType_GetModuleState(Py_TYPE(self)) only works if self type is exactly BZ2Decompressor. Otherwise, state is NULL or is the state of another module.

Same remark for _bz2_BZ2Compressor___init__().